man/mysqld_safe.1
'\" t
.\"     Title: mysqld_safe
.\"    Author: [FIXME: author] [see http://docbook.sf.net/el/author]
.\" Generator: DocBook XSL Stylesheets v1.79.1 <http://docbook.sf.net/>
.\"      Date: 06/25/2026
.\"    Manual: MySQL Database System
.\"    Source: MySQL 9.7
.\"  Language: English
.\"
.TH "MYSQLD_SAFE" "1" "06/25/2026" "MySQL 9\&.7" "MySQL Database System"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
mysqld_safe \- MySQL server startup script
.SH "SYNOPSIS"
.HP \w'\fBmysqld_safe\ \fR\fB\fIoptions\fR\fR\ 'u
\fBmysqld_safe \fR\fB\fIoptions\fR\fR
.SH "DESCRIPTION"
.PP
\fBmysqld_safe\fR
is the recommended way to start a
\fBmysqld\fR
server on Unix\&.
\fBmysqld_safe\fR
adds some safety features such as restarting the server when an error occurs and logging runtime information to an error log\&. A description of error logging is given later in this section\&.
.if n \{\
.sp
.\}
.RS 4
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBNote\fR
.ps -1
.br
.PP
For some Linux platforms, MySQL installation from RPM or Debian packages includes systemd support for managing MySQL server startup and shutdown\&. On these platforms,
\fBmysqld_safe\fR
is not installed because it is unnecessary\&. For more information, see
Section\ \&2.5.9, \(lqManaging MySQL Server with systemd\(rq\&.
.PP
One implication of the non\-use of
\fBmysqld_safe\fR
on platforms that use systemd for server management is that use of
[mysqld_safe]
or
[safe_mysqld]
sections in option files is not supported and might lead to unexpected behavior\&.
.sp .5v
.RE
.PP
\fBmysqld_safe\fR
tries to start an executable named
\fBmysqld\fR\&. To override the default behavior and specify explicitly the name of the server you want to run, specify a
\fB\-\-mysqld\fR
or
\fB\-\-mysqld\-version\fR
option to
\fBmysqld_safe\fR\&. You can also use
\fB\-\-ledir\fR
to indicate the directory where
\fBmysqld_safe\fR
should look for the server\&.
.PP
Many of the options to
\fBmysqld_safe\fR
are the same as the options to
\fBmysqld\fR\&. See
Section\ \&7.1.7, \(lqServer Command Options\(rq\&.
.PP
Options unknown to
\fBmysqld_safe\fR
are passed to
\fBmysqld\fR
if they are specified on the command line, but ignored if they are specified in the
[mysqld_safe]
group of an option file\&. See
Section\ \&6.2.2.2, \(lqUsing Option Files\(rq\&.
.PP
\fBmysqld_safe\fR
reads all options from the
[mysqld],
[server], and
[mysqld_safe]
sections in option files\&. For example, if you specify a
[mysqld]
section like this,
\fBmysqld_safe\fR
finds and uses the
\fB\-\-log\-error\fR
option:
.sp
.if n \{\
.RS 4
.\}
.nf
[mysqld]
log\-error=error\&.log
.fi
.if n \{\
.RE
.\}
.PP
For backward compatibility,
\fBmysqld_safe\fR
also reads
[safe_mysqld]
sections, but to be current you should rename such sections to
[mysqld_safe]\&.
.PP
\fBmysqld_safe\fR
accepts options on the command line and in option files, as described in the following table\&. For information about option files used by MySQL programs, see
Section\ \&6.2.2.2, \(lqUsing Option Files\(rq\&.
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-help\fR
.TS
allbox tab(:);
lB l.
T{
Command-Line Format
T}:T{
--help
T}
.TE
.sp 1
Display a help message and exit\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-basedir=\fR\fB\fIdir_name\fR\fR
.TS
allbox tab(:);
lB l
lB l.
T{
Command-Line Format
T}:T{
--basedir=dir_name
T}
T{
Type
T}:T{
Directory name
T}
.TE
.sp 1
The path to the MySQL installation directory\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-core\-file\-size=\fR\fB\fIsize\fR\fR
.TS
allbox tab(:);
lB l
lB l.
T{
Command-Line Format
T}:T{
--core-file-size=size
T}
T{
Type
T}:T{
String
T}
.TE
.sp 1
The size of the core file that
\fBmysqld\fR
should be able to create\&. The option value is passed to
\fBulimit \-c\fR\&.
.if n \{\
.sp
.\}
.RS 4
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBNote\fR
.ps -1
.br
The
innodb_buffer_pool_in_core_file
variable can be used to reduce the size of core files on operating systems that support it\&. For more information, see
Section\ \&17.8.3.7, \(lqExcluding or Including Buffer Pool Pages from Core Files\(rq\&.
.sp .5v
.RE
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-datadir=\fR\fB\fIdir_name\fR\fR
.TS
allbox tab(:);
lB l
lB l.
T{
Command-Line Format
T}:T{
--datadir=dir_name
T}
T{
Type
T}:T{
Directory name
T}
.TE
.sp 1
The path to the data directory\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-defaults\-extra\-file=\fR\fB\fIfile_name\fR\fR
.TS
allbox tab(:);
lB l
lB l.
T{
Command-Line Format
T}:T{
--defaults-extra-file=file_name
T}
T{
Type
T}:T{
File name
T}
.TE
.sp 1
Read this option file in addition to the usual option files\&. If the file does not exist or is otherwise inaccessible, the server exits with an error\&. If
\fIfile_name\fR
is not an absolute path name, it is interpreted relative to the current directory\&. This must be the first option on the command line if it is used\&.
.sp
For additional information about this and other option\-file options, see
Section\ \&6.2.2.3, \(lqCommand-Line Options that Affect Option-File Handling\(rq\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-defaults\-file=\fR\fB\fIfile_name\fR\fR
.TS
allbox tab(:);
lB l
lB l.
T{
Command-Line Format
T}:T{
--defaults-file=file_name
T}
T{
Type
T}:T{
File name
T}
.TE
.sp 1
Use only the given option file\&. If the file does not exist or is otherwise inaccessible, the server exits with an error\&. If
\fIfile_name\fR
is not an absolute path name, it is interpreted relative to the current directory\&. This must be the first option on the command line if it is used\&.
.sp
For additional information about this and other option\-file options, see
Section\ \&6.2.2.3, \(lqCommand-Line Options that Affect Option-File Handling\(rq\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-ledir=\fR\fB\fIdir_name\fR\fR
.TS
allbox tab(:);
lB l
lB l.
T{
Command-Line Format
T}:T{
--ledir=dir_name
T}
T{
Type
T}:T{
Directory name
T}
.TE
.sp 1
If
\fBmysqld_safe\fR
cannot find the server, use this option to indicate the path name to the directory where the server is located\&.
.sp
This option is accepted only on the command line, not in option files\&. On platforms that use systemd, the value can be specified in the value of
MYSQLD_OPTS\&. See
Section\ \&2.5.9, \(lqManaging MySQL Server with systemd\(rq\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-log\-error=\fR\fB\fIfile_name\fR\fR
.TS
allbox tab(:);
lB l
lB l.
T{
Command-Line Format
T}:T{
--log-error=file_name
T}
T{
Type
T}:T{
File name
T}
.TE
.sp 1
Write the error log to the given file\&. See
Section\ \&7.4.2, \(lqThe Error Log\(rq\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-mysqld\-safe\-log\-timestamps\fR
.TS
allbox tab(:);
lB l
lB l
lB l
lB l.
T{
Command-Line Format
T}:T{
--mysqld-safe-log-timestamps=type
T}
T{
Type
T}:T{
Enumeration
T}
T{
Default Value
T}:T{
utc
T}
T{
Valid Values
T}:T{
.PP
system
.PP
hyphen
.PP
legacy
T}
.TE
.sp 1
This option controls the format for timestamps in log output produced by
\fBmysqld_safe\fR\&. The following list describes the permitted values\&. For any other value,
\fBmysqld_safe\fR
logs a warning and uses
UTC
format\&.
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
UTC,
utc
.sp
ISO 8601 UTC format (same as
\fB\-\-log_timestamps=UTC\fR
for the server)\&. This is the default\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
SYSTEM,
system
.sp
ISO 8601 local time format (same as
\fB\-\-log_timestamps=SYSTEM\fR
for the server)\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
HYPHEN,
hyphen
.sp
\fIYY\-MM\-DD h:mm:ss\fR
format, as in
\fBmysqld_safe\fR
for MySQL 5\&.6\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
LEGACY,
legacy
.sp
\fIYYMMDD hh:mm:ss\fR
format, as in
\fBmysqld_safe\fR
prior to MySQL 5\&.6\&.
.RE
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-malloc\-lib=[\fR\fB\fIlib_name\fR\fR\fB]\fR
.TS
allbox tab(:);
lB l
lB l.
T{
Command-Line Format
T}:T{

ChangeLog
2010-03-18	The InnoDB Team

	* CMakeLists.txt:
	Fix Bug#52102 InnoDB Plugin shows performance drop compared to
	InnoDB (Windows)

2010-03-18	The InnoDB Team

	* buf0buf.ic:
	When comparing the time of the first access to a block against
	innodb_old_blocks_time, use 32-bit arithmetics. The comparison was
	incorrect on 64-bit systems.

2010-03-11	The InnoDB Team

	* buf0buf.h, buf0buf.ic:
	Fix and clarify the latching of some buf_block_t members.
	Note that check_index_page_at_flush is not protected by any mutex.
	Note and assert that lock_hash_val is protected by the rw-latch.

2010-03-10	The InnoDB Team

	* trx/trx0sys.c:
	Fix Bug#51653 outdated reference to set-variable

2010-03-10	The InnoDB Team

	* handler/ha_innodb.cc, mysql-test/innodb_bug21704.result,
	mysql-test/innodb_bug47621.result, mysql-test/innodb_bug47621.test:
	Fix Bug#47621 MySQL and InnoDB data dictionaries will become out of
	sync when renaming columns

2010-03-10	The InnoDB Team

	* handler/ha_innodb.cc:
	Fix Bug#51356 Many Valgrind errors in error messages
	with concurrent DDL

2010-03-10	The InnoDB Team

	* handler/ha_innodb.cc, handler/handler0alter.cc,
	mysql-test/innodb_bug51378.result, mysql-test/innodb_bug51378.test:
	Fix Bug#51378 Init 'ref_length' to correct value, in case an out
	of bound MySQL primary_key

2010-03-10	The InnoDB Team

	* log/log0recv.c:
	Remove a bogus assertion about page numbers exceeding 0x90000000
	in the redo log. Abort when encountering a corrupted redo log
	record, unless innodb_force_recovery is set.

2010-03-09	The InnoDB Team

	* handler/ha_innodb.cc:
	Make SHOW ENGINE INNODB MUTEX STATUS display SUM(os_waits)
	for the buffer pool block mutexes and locks.

2010-03-08	The InnoDB Team

	* fil/fil0fil.c:
	Fix ALTER TABLE ... IMPORT TABLESPACE of compressed tables.

2010-03-03	The InnoDB Team

	* handler/handler0alter.cc, innodb-index.result, innodb-index.test,
	innodb.result, innodb.test:
	Disallow a duplicate index name when creating an index.

2010-02-11	The InnoDB Team

	* include/mem0mem.h, include/mem0mem.ic, mem/mem0mem.c:
	Fix Bug#49535 Available memory check slows down crash
	recovery tens of times

2010-02-09	The InnoDB Team

	* buf/buf0buf.c:
	Fix Bug#38901 InnoDB logs error repeatedly when trying to load
	page into buffer pool

2010-02-09	The InnoDB Team

	* srv/srv0srv.c:
	Let the master thread sleep if the amount of work to be done is
	calibrated as taking less than a second.

2010-02-04	The InnoDB Team

	* btr/btr0btr.c, btr/btr0cur.c, btr/btr0pcur.c, buf/buf0buf.c,
	include/btr0btr.h, include/btr0cur.h, include/btr0pcur.h,
	include/btr0pcur.ic, include/buf0buf.h, row/row0ins.c, row/row0sel.c:
	Pass the file name and line number of the caller of the
	b-tree cursor functions to the buffer pool requests, in order
	to make the latch diagnostics more accurate.

2010-02-03	The InnoDB Team

	* lock/lock0lock.c:
	Fix Bug#49001 SHOW INNODB STATUS deadlock info incorrect
	when deadlock detection aborts

2010-02-03	The InnoDB Team

	* buf/buf0lru.c:
	Fix Bug#35077 Very slow DROP TABLE (ALTER TABLE, OPTIMIZE TABLE)
	on compressed tables

2010-02-03	The InnoDB Team

	* handler/ha_innodb.cc, include/row0mysql.h, row/row0mysql.c:
	Clean up CHECK TABLE error handling.

2010-02-01	The InnoDB Team

	* handler/ha_innodb.cc, mysql-test/innodb-autoinc.test,
	mysql-test/innodb-autoinc.result,
	mysql-test/innodb-autoinc-44030.test,
	mysql-test/innodb-autoinc-44030.result:
	Fix Bug#49497 Error 1467 (ER_AUTOINC_READ_FAILED) on inserting
	a negative value

2010-01-27	The InnoDB Team

	* include/row0mysql.h, log/log0recv.c, row/row0mysql.c:
	Drop temporary tables at startup.
	This addresses the third aspect of
	Bug#41609 Crash recovery does not work for InnoDB temporary tables.

2010-01-21	The InnoDB Team

	* buf/buf0buf.c:
	Do not merge buffered inserts to compressed pages before
	the redo log has been applied in crash recovery.

2010-01-13	The InnoDB Team

	* row/row0sel.c:
	On the READ UNCOMMITTED isolation level, do not attempt to access
	a clustered index record that has been marked for deletion. The
	built-in InnoDB in MySQL 5.1 and earlier would attempt to retrieve
	a previous version of the record in this case.

2010-01-13	The InnoDB Team

	* buf/buf0buf.c:
	When disabling the adaptive hash index, check the block state
	before checking block->is_hashed, because the latter may be
	uninitialized right after server startup.

2010-01-12	The InnoDB Team

	* handler/ha_innodb.cc, handler/ha_innodb.h:
	Fix Bug#46193 crash when accessing tables after enabling
	innodb_force_recovery option

2010-01-12	The InnoDB Team

	* row/row0mysql.c:

netware/mysqld_safe.c
/*
  Copyright (c) 2003 Novell, Inc. All Rights Reserved.

  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA
*/

#include <stdlib.h>
#include <stdio.h>
#include <netdb.h>
#include <sys/stat.h>
#include <monitor.h>
#include <strings.h>
#include <getopt.h>
#include <screen.h>
#include <dirent.h>

#include "my_config.h"
#include "my_manage.h"
#include "mysql_version.h"

/******************************************************************************

	global variables

******************************************************************************/
char autoclose;
char basedir[PATH_MAX];
char checktables;
char datadir[PATH_MAX];
char pid_file[PATH_MAX];
char address[PATH_MAX];
char port[PATH_MAX];
char err_log[PATH_MAX];
char safe_log[PATH_MAX];
char mysqld[PATH_MAX];
char hostname[PATH_MAX];
char default_option[PATH_MAX];

FILE *log_fd= NULL;

/******************************************************************************

	prototypes

******************************************************************************/

void usage(void);
void vlog(char *, va_list);
void log(char *, ...);
void start_defaults(int, char *[]);
void finish_defaults();
void read_defaults(arg_list_t *);
void parse_args(int, char *[]);
void get_options(int, char *[]);
void check_data_vol();
void check_setup();
void check_tables();
void mysql_start(int, char *[]);
void parse_setvar(char *arg);

/******************************************************************************

	functions

******************************************************************************/

/******************************************************************************

  usage()

  Show usage.

******************************************************************************/
void usage(void)
{
  // keep the screen up
  setscreenmode(SCR_NO_MODE);

  puts("\
\n\
usage: mysqld_safe [options]\n\
\n\
Program to start the MySQL daemon and restart it if it dies unexpectedly.\n\
All options, besides those listed below, are passed on to the MySQL daemon.\n\
\n\
options:\n\
\n\
--autoclose                 Automatically close the mysqld_safe screen.\n\
\n\
--check-tables              Check the tables before starting the MySQL daemon.\n\
\n\
--err-log=<file>            Send the MySQL daemon error output to <file>.\n\
\n\
--help                      Show this help information.\n\
\n\
--mysqld=<file>             Use the <file> MySQL daemon.\n\
\n\
  ");

  exit(-1);
}

/******************************************************************************

  vlog()

  Log the message.

******************************************************************************/
void vlog(char *format, va_list ap)
{
  vfprintf(stdout, format, ap);
  fflush(stdout);

  if (log_fd)
  {
    vfprintf(log_fd, format, ap);
    fflush(log_fd);
  }
}

/******************************************************************************

  log()

  Log the message.

******************************************************************************/
void log(char *format, ...)
{
  va_list ap;

  va_start(ap, format);

  vlog(format, ap);

  va_end(ap);
}

/******************************************************************************

	start_defaults()

	Start setting the defaults.

******************************************************************************/
void start_defaults(int argc, char *argv[])
{
  struct stat buf;
  int i;

  // default options
  static char *default_options[]=
  {
    "--no-defaults",
    "--defaults-file=",
    "--defaults-extra-file=",
    NULL
  };

  // autoclose
  autoclose= FALSE;

  // basedir
  get_basedir(argv[0], basedir);

  // check-tables
  checktables= FALSE;

  // hostname
  if (gethostname(hostname, PATH_MAX) < 0)
  {
    // default
    strcpy(hostname, "mysql");
  }

  // address
  snprintf(address, PATH_MAX, "0.0.0.0");

  // port
  snprintf(port, PATH_MAX, "%d", MYSQL_PORT);

  // default option
  default_option[0]= 0;
  for (i= 0; (argc > 1) && default_options[i]; i++)
  {
    if (!strnicmp(argv[1], default_options[i], strlen(default_options[i])))
    {
      strncpy(default_option, argv[1], PATH_MAX);
      break;
    }
  }

  // set after basedir is established
  datadir[0]= 0;
  pid_file[0]= 0;
  err_log[0]= 0;
  safe_log[0]= 0;
  mysqld[0]= 0;
}

/******************************************************************************

	finish_defaults()

	Finish settig the defaults.

******************************************************************************/
void finish_defaults()
{
  struct stat buf;
  int i;

  // datadir
  if (!datadir[0])
    snprintf(datadir, PATH_MAX, "%s/data", basedir);

  // pid-file
  if (!pid_file[0])
    snprintf(pid_file, PATH_MAX, "%s/%s.pid", datadir, hostname);

  // err-log
  if (!err_log[0])
    snprintf(err_log, PATH_MAX, "%s/%s.err", datadir, hostname);

  // safe-log
  if (!safe_log[0])
    snprintf(safe_log, PATH_MAX, "%s/%s.safe", datadir, hostname);

  // mysqld
  if (!mysqld[0])
    snprintf(mysqld, PATH_MAX, "%s/bin/mysqld-max", basedir);

  if (stat(mysqld, &buf))
  {
    snprintf(mysqld, PATH_MAX, "%s/bin/mysqld", basedir);
  }
}

/******************************************************************************

	read_defaults()

	Read the defaults.

******************************************************************************/
void read_defaults(arg_list_t *pal)
{
  arg_list_t al;
  char defaults_file[PATH_MAX];
  char mydefaults[PATH_MAX];
  char line[PATH_MAX];
  FILE *fp;

  // defaults output file
  snprintf(defaults_file, PATH_MAX, "%s/bin/defaults.out", basedir);
  remove(defaults_file);

  // mysqladmin file
  snprintf(mydefaults, PATH_MAX, "%s/bin/my_print_defaults", basedir);

  // args
  init_args(&al);
  add_arg(&al, mydefaults);
  if (default_option[0])
    add_arg(&al, default_option);
  add_arg(&al, "mysqld");
  add_arg(&al, "server");
  add_arg(&al, "mysqld_safe");
  add_arg(&al, "safe_mysqld");

  spawn(mydefaults, &al, TRUE, NULL, defaults_file, NULL);

  free_args(&al);

  // gather defaults
  if ((fp= fopen(defaults_file, "r")) != NULL)
  {
    while (fgets(line, PATH_MAX, fp))
    {
      char *p;

      // remove end-of-line character
      if ((p= strrchr(line, '\n')) != NULL)
	*p= '\0';

      // add the option as an argument
      add_arg(pal, line);
    }

    fclose(fp);
  }

  // remove file
  remove(defaults_file);
}

/******************************************************************************

	parse_args()

	Get the options.

******************************************************************************/
void parse_args(int argc, char *argv[])
{
  int index= 0;
  int c;

  // parse options
  enum opts
  {
    OPT_BASEDIR= 0xFF,
    OPT_DATADIR,
    OPT_PID_FILE,
    OPT_BIND_ADDRESS,
    OPT_PORT,
    OPT_ERR_LOG,
    OPT_SAFE_LOG,
    OPT_MYSQLD,
    OPT_HELP,
    OPT_SETVAR
  };

  static struct option options[]=
  {
    {"autoclose", no_argument, &autoclose, TRUE},
    {"basedir", required_argument, 0, OPT_BASEDIR},
    {"check-tables", no_argument, &checktables, TRUE},
    {"datadir", required_argument, 0, OPT_DATADIR},
    {"pid-file", required_argument, 0, OPT_PID_FILE},
    {"bind-address", required_argument, 0, OPT_BIND_ADDRESS},
    {"port", required_argument, 0, OPT_PORT},
    {"err-log", required_argument, 0, OPT_ERR_LOG},
    {"safe-log", required_argument, 0, OPT_SAFE_LOG},
    {"mysqld", required_argument, 0, OPT_MYSQLD},
    {"help", no_argument, 0, OPT_HELP},
    {"set-variable", required_argument, 0, OPT_SETVAR},
    {0, 0, 0, 0}
  };

  // we have to reset getopt_long because we use it multiple times
  optind= 1;

  // turn off error reporting
  opterr= 0;

  while ((c= getopt_long(argc, argv, "b:h:P:", options, &index)) >= 0)
  {
    switch (c) {
    case OPT_BASEDIR:
    case 'b':
      strcpy(basedir, optarg);
      break;

    case OPT_DATADIR:
    case 'h':
      strcpy(datadir, optarg);
      break;

    case OPT_PID_FILE:
      strcpy(pid_file, optarg);
      break;

    case OPT_BIND_ADDRESS:
      strcpy(address, optarg);
      break;

    case OPT_PORT:
    case 'P':
      strcpy(port, optarg);
      break;

    case OPT_ERR_LOG:
      strcpy(err_log, optarg);
      break;

    case OPT_SAFE_LOG:
      strcpy(safe_log, optarg);
      break;

    case OPT_MYSQLD:
      strcpy(mysqld, optarg);
      break;

    case OPT_SETVAR:
      parse_setvar(optarg);
      break;

    case OPT_HELP:
      usage();
      break;

    default:
      // ignore
      break;
    }
  }
}

/*
  parse_setvar(char *arg)
  Pasrsing for port just to display the port num on the mysqld_safe screen
*/
void parse_setvar(char *arg)
{
  char *pos;

  if ((pos= strindex(arg, "port")))
  {
    for (; *pos && *pos != '='; pos++);
    if (*pos)
      strcpy(port, pos + 1);
  }
}

/******************************************************************************



/******************************************************************************

	get_options()

	Get the options.

******************************************************************************/
void get_options(int argc, char *argv[])
{
  arg_list_t al;

  // start defaults
  start_defaults(argc, argv);

  // default file arguments
  init_args(&al);
  add_arg(&al, "ignore");
  read_defaults(&al);
  parse_args(al.argc, al.argv);
  free_args(&al);

  // command-line arguments
  parse_args(argc, argv);

  // finish defaults
  finish_defaults();
}

/******************************************************************************

	check_data_vol()

	Check the database volume.

******************************************************************************/
void check_data_vol()
{
  // warn if the data is on a Traditional volume
  struct volume_info vol;
  char buff[PATH_MAX];
  char *p;

  // clear struct
  memset(&vol, 0, sizeof(vol));

  // find volume name
  strcpy(buff, datadir);
  if (p= strchr(buff, ':'))
  {
    // terminate after volume name
    *p= 0;
  }
  else
  {
    // assume SYS volume
    strcpy(buff, "SYS");
  }

  // retrieve information
  netware_vol_info_from_name(&vol, buff);

  if ((vol.flags & VOL_NSS_PRESENT) == 0)
  {
    log("Error: Either the data directory does not exist or is not on an NSS volume!\n\n");
    exit(-1);
  }
}

/******************************************************************************

	check_setup()

	Check the current setup.

******************************************************************************/
void check_setup()
{
  struct stat info;
  char temp[PATH_MAX];

  // remove any current pid_file
  if (!stat(pid_file, &info) && (remove(pid_file) < 0))
  {
    log("ERROR: Unable to remove current pid file!\n\n");
    exit(-1);
  }

  // check the data volume
  check_data_vol();

  // check for a database
  snprintf(temp, PATH_MAX, "%s/mysql/host.frm", datadir);
  if (stat(temp, &info))
  {
    log("ERROR: No database found in the data directory!\n\n");
    exit(-1);
  }
}

/******************************************************************************

	check_tables()

	Check the database tables.

******************************************************************************/
void check_tables()
{
  arg_list_t al;
  char mycheck[PATH_MAX];
  char table[PATH_MAX];
  char db[PATH_MAX];
  DIR *datadir_entry, *db_entry, *table_entry;

  // status
  log("checking tables...\n");

  // list databases
  if ((datadir_entry= opendir(datadir)) == NULL)
  {
    return;
  }

  while ((db_entry= readdir(datadir_entry)) != NULL)
  {
    if (db_entry->d_name[0] == '.')
    {
      // Skip
    }
    else if (S_ISDIR(db_entry->d_type))
    {
      // create long db name
      snprintf(db, PATH_MAX, "%s/%s", datadir, db_entry->d_name);

      // list tables
      if ((db_entry= opendir(db)) == NULL)
      {
	continue;
      }

      while ((table_entry= readdir(db_entry)) != NULL)
      {
	// create long table name
	snprintf(table, PATH_MAX, "%s/%s", db, strlwr(table_entry->d_name));

	if (strindex(table, ".myi"))
	{
	  // ** myisamchk

	  // mysqladmin file
	  snprintf(mycheck, PATH_MAX, "%s/bin/myisamchk", basedir);

	  // args
	  init_args(&al);
	  add_arg(&al, mycheck);
	  add_arg(&al, "--silent");
	  add_arg(&al, "--force");
	  add_arg(&al, "--fast");
	  add_arg(&al, "--medium-check");
	  add_arg(&al, "-O");
	  add_arg(&al, "key_buffer=64M");
	  add_arg(&al, "-O");
	  add_arg(&al, "sort_buffer=64M");
	  add_arg(&al, table);

	  spawn(mycheck, &al, TRUE, NULL, NULL, NULL);

	  free_args(&al);
	}
	else if (strindex(table, ".ism"))
	{
	  // ** isamchk

	  // mysqladmin file
	  snprintf(mycheck, PATH_MAX, "%s/bin/isamchk", basedir);

	  // args
	  init_args(&al);
	  add_arg(&al, mycheck);
	  add_arg(&al, "--silent");
	  add_arg(&al, "--force");
	  add_arg(&al, "-O");
	  add_arg(&al, "sort_buffer=64M");
	  add_arg(&al, table);

	  spawn(mycheck, &al, TRUE, NULL, NULL, NULL);

	  free_args(&al);
	}
      }
    }
  }
}

/******************************************************************************

	mysql_start()

	Start the mysql server.

******************************************************************************/
void mysql_start(int argc, char *argv[])
{
  arg_list_t al;
  int i, j, err;
  struct stat info;
  time_t cal;
  struct tm lt;
  char stamp[PATH_MAX];
  char skip;

  // private options
  static char *private_options[]=
  {
    "--autoclose",
    "--check-tables",
    "--help",
    "--err-log=",
    "--mysqld=",
    NULL
  };

  // args
  init_args(&al);
  add_arg(&al, "%s", mysqld);

  // parent args
  for (i= 1; i < argc; i++)
  {
    skip= FALSE;

    // skip private arguments
    for (j= 0; private_options[j]; j++)
    {
      if (!strnicmp(argv[i], private_options[j], strlen(private_options[j])))
      {
	skip= TRUE;
	break;
      }
    }

    if (!skip)
    {
      add_arg(&al, "%s", argv[i]);
    }
  }
  // spawn
  do
  {
    // check the database tables
    if (checktables)
      check_tables();

    // status
    time(&cal);
    localtime_r(&cal, &lt);
    strftime(stamp, PATH_MAX, "%d %b %Y %H:%M:%S", &lt);
    log("mysql started    : %s\n", stamp);

    // spawn mysqld
    spawn(mysqld, &al, TRUE, NULL, NULL, err_log);
  }
  while (!stat(pid_file, &info));

  // status
  time(&cal);
  localtime_r(&cal, &lt);
  strftime(stamp, PATH_MAX, "%d %b %Y %H:%M:%S", &lt);
  log("mysql stopped    : %s\n\n", stamp);

  // free args
  free_args(&al);
}

/******************************************************************************

	main()

******************************************************************************/
int main(int argc, char **argv)
{
  char temp[PATH_MAX];

  // get the options
  get_options(argc, argv);

  // keep the screen up
  if (!autoclose)
    setscreenmode(SCR_NO_MODE);

  // create log file
  log_fd= fopen(safe_log, "w+");

  // header
  log("MySQL Server %s, for %s (%s)\n\n", VERSION, SYSTEM_TYPE, MACHINE_TYPE);

  // status
  log("address          : %s\n", address);
  log("port             : %s\n", port);
  log("daemon           : %s\n", mysqld);
  log("base directory   : %s\n", basedir);
  log("data directory   : %s\n", datadir);
  log("pid file         : %s\n", pid_file);
  log("error file       : %s\n", err_log);
  log("log file         : %s\n", safe_log);
  log("\n");

  // check setup
  check_setup();

  // start the MySQL server
  mysql_start(argc, argv);

  // close log file
  if (log_fd)
    fclose(log_fd);

  return 0;
}
